guineaPigFacts/src/main/java/com/factsdemo/guineaPigFacts/controllers/UserController.java

```diff
@@ -29,21 +30,35 @@
     @Autowired
     UserService userService;
 
+    /**
+     * Parses User so the password is not also returned as an HTTP response. Uses all other parts of
+     * User to create the object to be sent.
+     * @param user
+     * @return JSONObject
+     */
+    private JSONObject getSafeUser(User user) {
+        JSONObject safeUser = new JSONObject();
```

Comment:
JSONObject is an interesting class.  Normally, I would say to use a custom class similar to the User class so that you can keep the benefits of compile-time type-checking.  However, in this case, I think we'll probably only be using the SafeUser as a value in an HTTP response body.  I can't really think of other use-cases in the app so this choice of data structure is fine.

With that in mind, a better way to do this would be to use a Filter.  Filters are useful for performing logic on all requests or responses that match a specific criteria.  For example, for this, you could create a Filter that automatically strips the password and any other sensitive information from any response that comes from the UserController.  This is nice because it allows you to add new methods to your controller and any response object will automatically be safe.  You don't have to remember to explicitly strip out sensitive information.  Always try to automate things away rather than having to rely on a checklist or your memory when developing.

You can see this website for a tutorial on how to do it[https://www.techiedelight.com/add-custom-header-to-all-responses-spring-boot/]( https://www.techiedelight.com/add-custom-header-to-all-responses-spring-boot/).

Reply:
Okay, so from what I can see I need to do the following steps to create the Filter:
- I need to create an HttpServletResponseWrapper which is where the response is changed.
- I need a custom OutputStream wrapper to be used in the HttpServletResponseWrapper to actually transform the responses.
- Then, I call the wrapped response in doFilter() which will be called with the client request post&pre processing 
Where I got the idea: https://stackoverflow.com/questions/5634477/how-to-write-response-filter

Reply:
Yeah that sounds about right.  Here is another step by step example with code typed out that also does a really good example of explaining what each thing does.  If you have any questions, about how this all works, let me know.

[https://stackoverflow.com/a/25155407](https://stackoverflow.com/a/25155407)

Nice work figuring it out!  This actually got pretty complicated :)